headerfs/js.go

```diff
@@ -0,0 +1,402 @@
+//go:build js && wasm
+
+package headerfs
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall/js"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/linden/indexeddb"
+)
+
+const headersStore = "headers"
+
+// headerStore combines a IndexedDB store of headers within a flat file in addition
+// to a database which indexes that flat file. Together, these two abstractions
+// can be used in order to build an indexed header store for any type of
+// "header" as it deals only with raw bytes, and leaves it to a higher layer to
+// interpret those raw bytes accordingly.
+//
+// TODO(roasbeef): quickcheck coverage.
+type headerStore struct {
+	mtx sync.RWMutex // nolint:structcheck // false positive because used as embedded struct only
+
+	idb *indexeddb.DB
+
+	*headerIndex
+}
+
+// readHeader reads a full block header from the flat-file. The header read is
+// determined by the hight value.
+func (h *headerStore) readHeader(height uint32) (wire.BlockHeader, error) {
+	// Create a new read/write transaction, scoped to headers.
+	tx, err := h.idb.NewTransaction([]string{headersStore}, indexeddb.ReadWriteMode)
+	if err != nil {
+		return wire.BlockHeader{}, err
+	}
+
+	// Get the headers store.
+	hdrs := tx.Store(headersStore)
+
+	// Get the raw header.
+	val, err := hdrs.Get(height)
+
+	// check if the value was not found as that case is handled differently.
+	if errors.Is(err, indexeddb.ErrValueNotFound) {
+		return wire.BlockHeader{}, &ErrHeaderNotFound{err}
+	}
+
+	if err != nil {
+		return wire.BlockHeader{}, err
+	}
+
+	// Ensure the value is a string.
+	if val.Type() != js.TypeString {
+		return wire.BlockHeader{}, fmt.Errorf("unexpected type: %s", val.Type())
+	}
+
+	var header wire.BlockHeader
+
+	// Unquote the value.
+	unquote, err := strconv.Unquote(val.String())
+	if err != nil {
+		return wire.BlockHeader{}, err
+	}
+
+	// Create a new reader from the raw header.
+	reader := strings.NewReader(unquote)
+
+	// Finally, decode the  bytes into a proper bitcoin header.
+	err = header.Deserialize(reader)
+
+	return header, err
+}
+
+func (h *headerStore) height() (uint32, bool, error) {
+	// Create a new read-only transaction. Scoped to headers.
+	tx, err := h.idb.NewTransaction([]string{headersStore}, indexeddb.ReadMode)
+	if err != nil {
+		return 0, false, err
+	}
+
+	// Get the headers store.
+	hdrs := tx.Store(headersStore)
+
+	// Count the amount of headers.
+	count, err := hdrs.Count()
+
+	// Fallback to the zero value if no value is found.
+	if errors.Is(err, indexeddb.ErrValueNotFound) || count == 0 {
+		return 0, true, nil
+	}
+
+	if err != nil {
+		return 0, false, err
+	}
+
+	// Subtract one as the block height does not include the genesis block.
+	return uint32(count - 1), false, nil
+}
+
+func (h *headerStore) singleTruncate() error {
+	// Get the current height.
+	height, genesis, err := h.height()
```

Comment:
Need to check out the underlying implementation, but: should the height be obtained _after_ the transaction is created? Otherwise, two writers may start the transaction with an inconsistent view.